cmd/snap-confine/xauth.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "../libsnap-confine-private/secure-getenv.h"
+
+#include "xauth.h"
+
+#include <memory.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include <linux/limits.h>
+
+static char *xauth_data = NULL;
+static long xauth_data_length = 0;
+
+void sc_xauth_load_from_env(void)
+{
+	const char *xauth_path = getenv("XAUTHORITY");
+	if (!xauth_path)
+		return;
+
+	FILE *f = fopen(xauth_path, "rb");
+	if (!f)
+		return;
+
+	fseek(f, 0, SEEK_END);
```

Comment:
This could be used to consume arbitrary amount of data. I think you want to use a fixed buffer of sane size (check how big the auth cookie is) and not use more.